File: samples/Sample/PushActionsVC.m
/*
 * Licensed Materials - Property of IBM
 *
 * 5725E28, 5725I03
 *
 * © Copyright IBM Corp. 2011, 2017
 * US Government Users Restricted Rights - Use, duplication or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 */
#import "PushActionsVC.h"
#import "EditCell.h"
#import <IBMMobilePush/IBMMobilePush.h>
#import <objc/runtime.h>

@interface PushActionsVC ()
@property UIColor * standardTextColor;
@property UITextField * standardValue;
@property UITextField * customType;
@property UITextField * customValue;
@property id observer;
@end

@implementation PushActionsVC

#pragma mark UITableViewDataSource
-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSUserDefaults * defaults = [NSUserDefaults standardUserDefaults];
    if(indexPath.item==0 && indexPath.section==0)
    {
        UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:@"view" forIndexPath:indexPath];
        cell.textLabel.text=@"Type";
        cell.detailTextLabel.text=[defaults stringForKey:@"standardType"];
        return cell;
    }
    else if(indexPath.item==0 || indexPath.item==1)
    {
        EditCell * cell = [tableView dequeueReusableCellWithIdentifier:@"edit"];
        UITextField * detail = cell.editField;
        detail.delegate=self;
        NSString * key = nil;
        if(indexPath.item==0)
        {
            cell.textField.text=@"Type";
            key = @"customType";
            self.customType = detail;
        }
        else if(indexPath.item==1)
        {
            cell.textField.text=@"Value";
            if(indexPath.section==0)
            {
                NSString * type = [defaults valueForKey:@"standardType"];
                if([type isEqual:@"url"])
                    key = @"standardUrlValue";
                else if([type isEqual:@"dial"])
                    key = @"standardDialValue";
                self.standardValue = detail;
            }
            else
            {
                key = @"customValue";
                self.customValue = detail;
            }
        }
        if(key)
        {
            detail.text = [defaults stringForKey:key];
            detail.enabled=TRUE;
        }
        else
        {
            detail.text=@"";
            detail.enabled=FALSE;
        }
        objc_setAssociatedObject(detail, @"key", key, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
        objc_setAssociatedObject(detail, @"indexPath", indexPath, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
        return cell;
    }
    else if(indexPath.item==2)
    {
        UITableViewCell * cell = [tableView dequeueReusableCellWithIdentifier:@"longview" forIndexPath:indexPath];
        cell.textLabel.text=@"JSON";
        NSString * type=nil;
        NSString * value = nil;
        if(indexPath.section==0)
        {
            type=[defaults stringForKey:@"standardType"];

            if([[defaults valueForKey:@"standardType"] isEqual:@"url"])
                value = [defaults stringForKey:@"standardUrlValue"];
            if([[defaults valueForKey:@"standardType"] isEqual:@"dial"])
                value = [defaults stringForKey:@"standardDialValue"];
        }
        else
        {
            type = [defaults stringForKey:@"customType"];
            value = [defaults stringForKey:@"customValue"];
        }
        
        if(value)
        {
            NSData * valueData = [value dataUsingEncoding:NSUTF8StringEncoding];
            NSError * error = nil;
            value = [NSJSONSerialization JSONObjectWithData:valueData options:NSJSONReadingAllowFragments error:&error];
            if(error)
            {
                [[[MCESdk.sharedInstance.alertViewClass alloc] initWithTitle:@"Value entry invalid" message:@"Could not serialize json. Is it valid?" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil]show];
            }
        }
        
        NSDictionary * jsonDict = value ? @{@"type": type, @"value": value} : @{@"type": type};
        
        NSError * jsonError = nil;
        NSData * jsonData = [NSJSONSerialization dataWithJSONObject:jsonDict options:0 error:&jsonError];
        cell.detailTextLabel.text = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
        cell.detailTextLabel.text = [cell.detailTextLabel.text stringByReplacingOccurrencesOfString:@"\\\"" withString:@"\""];
        cell.detailTextLabel.text = [cell.detailTextLabel.text stringByReplacingOccurrencesOfString:@"\\/" withString:@"/"];        return cell;
    }
    return nil;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 3;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 2;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    if(section==0)
        return @"Standard Action";
    return @"Custom Action";
}

-(NSString *)tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section
{
    if(section==0)
        return @"The action opens a dialer to call the number that is specified in the \"value\" key";
    return @"Please look at the application code for the implementation of custom actions (customAction). You can use that code as a starting point to create more custom actions. For more details, please see http://ibm.co/1Fp0OEQ";
}

#pragma mark UITableViewDelegate

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:TRUE];
    if(indexPath.item==1 && indexPath.section==0)
    {
        [self.standardValue becomeFirstResponder];
    }
    else if(indexPath.item==0 && indexPath.section==1)
    {
        [self.customType becomeFirstResponder];
    }
    else if(indexPath.item==1 && indexPath.section==1)
    {
        [self.customValue becomeFirstResponder];
    }
    else
    {
        [self.standardValue resignFirstResponder];
        [self.customType resignFirstResponder];
        [self.customValue resignFirstResponder];
    }
    
    if(indexPath.item==0 && indexPath.section==0)
    {
        UIActionSheet * sheet = [[UIActionSheet alloc] initWithTitle:@"Select Standard Action" delegate:self cancelButtonTitle:@"Cancel" destructiveButtonTitle:nil otherButtonTitles: @"dial", @"url", @"openApp", nil];
        objc_setAssociatedObject(sheet, @"indexPath", indexPath, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
        [sheet showInView: self.tableView];
    }
}
#pragma mark UITextFieldDelegate
- (void)textFieldDidEndEditing:(UITextField *)textField
{
    NSString * key = objc_getAssociatedObject(textField, @"key");
    NSIndexPath * indexPath = objc_getAssociatedObject(textField, @"indexPath");
    NSIndexPath * jsonIndexPath = [NSIndexPath indexPathForItem:2 inSection:indexPath.section];
    [[NSUserDefaults standardUserDefaults] setObject:textField.text forKey:key];
    [self.tableView reloadRowsAtIndexPaths:@[jsonIndexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [textField resignFirstResponder];
    return YES;
}

#pragma mark UIActionSheetDelegate
- (void)actionSheet:(UIActionSheet *)actionSheet clickedButtonAtIndex:(NSInteger)buttonIndex
{
    if(buttonIndex==0)
        [[NSUserDefaults standardUserDefaults] setObject:@"dial" forKey:@"standardType"];
    if(buttonIndex==1)
        [[NSUserDefaults standardUserDefaults] setObject:@"url" forKey:@"standardType"];
    if(buttonIndex==2)
        [[NSUserDefaults standardUserDefaults] setObject:@"openApp" forKey:@"standardType"];

    NSIndexPath * indexPath = objc_getAssociatedObject(actionSheet, @"indexPath");
    NSIndexPath * jsonIndexPath = [NSIndexPath indexPathForItem:2 inSection:indexPath.section];
    NSIndexPath * valueIndexPath = [NSIndexPath indexPathForItem:1 inSection:indexPath.section];
    [self.tableView reloadRowsAtIndexPaths:@[indexPath, jsonIndexPath, valueIndexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
}
@end

File: samples/SwiftSample/PushActionsVC.swift
/*
 * Licensed Materials - Property of IBM
 *
 * 5725E28, 5725I03
 *
 * © Copyright IBM Corp. 2015, 2016
 * US Government Users Restricted Rights - Use, duplication or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 */

import UIKit

enum PushActionSections: Int
{
    case standard
    case custom
    static let count: Int = 2
}

enum PushActionItems: Int
{
    case type
    case value
    case json
    static let count: Int = 3
}

enum PushActionTypes: Int
{
    case dial=1
    case url
    case openApp
}

class PushActionsVC : UITableViewController, UITextFieldDelegate, UIActionSheetDelegate
{
    @IBOutlet weak var standardValue: KeyedTextField?
    @IBOutlet weak var customType: KeyedTextField?
    @IBOutlet weak var customValue: KeyedTextField?
    var standardTextColor: UIColor?

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        let defaults = UserDefaults.standard
        
        if let pushActionSection = PushActionSections(rawValue: indexPath.section)
        {
            if let pushActionItem = PushActionItems(rawValue: indexPath.item)
            {
                switch(pushActionItem)
                {
                case .type:
                    if pushActionSection == .standard
                    {
                        let cell = tableView.dequeueReusableCell(withIdentifier: "view", for: indexPath as IndexPath)
                        cell.textLabel!.text = "Type"
                        cell.detailTextLabel!.text = defaults.string(forKey: "standardType")
                        return cell
                    }
                    else
                    {
                        let cell = tableView.dequeueReusableCell(withIdentifier: "edit", for: indexPath as IndexPath) as! EditCell
                        cell.textField!.text = "Type"
                        cell.editField!.text = defaults.string(forKey: "customType")
                        customType = cell.editField
                        customType!.indexPath = indexPath as IndexPath
                        customType!.key = "customType"
                        customType!.delegate = self
                        return cell
                    }
                    
                case .value:
                    
                    let cell = tableView.dequeueReusableCell(withIdentifier: "edit", for: indexPath as IndexPath) as! EditCell
                    cell.textField!.text = "Value"
                    if pushActionSection == .standard
                    {
                        standardValue = cell.editField!
                        standardValue!.delegate=self
                        if let type = defaults.string(forKey: "standardType")
                        {
                            if type == "url"
                            {
                                standardValue!.key = "standardUrlValue"
                            }
                            else if type == "dial"
                            {
                                standardValue!.key = "standardDialValue"
                            }
                        }
                        standardValue!.indexPath = indexPath as IndexPath
                        if let key = standardValue!.key
                        {
                            standardValue!.text = defaults.string(forKey: key)
                        }
                        else
                        {
                            standardValue!.text = ""
                        }
                    }
                    else
                    {
                        customValue = cell.editField!
                        customValue!.delegate=self
                        customValue!.key = "customValue"
                        customValue!.indexPath = indexPath as IndexPath
                        customValue!.text = defaults.string(forKey: customValue!.key!)
                    }
                    
                    return cell
                case .json:
                    let cell = tableView.dequeueReusableCell(withIdentifier: "longview", for: indexPath as IndexPath)
                    cell.textLabel!.text = "JSON"
                    var type: String?
                    var value: String?
                    if pushActionSection == .standard
                    {
                        type = defaults.string(forKey: "standardType")
                        if type == "url"
                        {
                            value = defaults.string(forKey: "standardUrlValue")
                        }
                        else if type == "dial"
                        {
                            value = defaults.string(forKey: "standardDialValue")
                        }
                    }
                    else
                    {
                        type = defaults.string(forKey: "customType")
                        value = defaults.string(forKey: "customValue")
                    }
                    
                    var jsonValue: Any?
                    if let v = value
                    {
                        let valueData = v.data(using: String.Encoding.utf8)
                        do
                        {
                            jsonValue = try JSONSerialization.jsonObject(with: valueData!, options: JSONSerialization.ReadingOptions.allowFragments )
                        }
                        catch _
                        {
                            UIAlertView.init(title: "Value entry invalid", message: "Could not serialize json. Is it valid?", delegate: nil, cancelButtonTitle: nil, otherButtonTitles: "OK").show()
                        }
                    }
                    
                    let jsonDict = ((jsonValue != nil) ? ["type": type!, "value": jsonValue!] : ["type": type!] ) as NSDictionary
                    
                    do
                    {
                        let data = try JSONSerialization.data(withJSONObject: jsonDict, options: JSONSerialization.WritingOptions())
                        var str = String(data: data, encoding: String.Encoding.utf8)
                        str = str?.replacingOccurrences(of: "\\\"", with: "\"")
                        str = str?.replacingOccurrences(of: "\\/", with: "/")
                        cell.detailTextLabel!.text = str
                    }
                    catch _
                    {
                        cell.detailTextLabel!.text = "Could not convert"
                    }
                    return cell
                }
            }
        }
        return UITableViewCell.init()
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath)
    {
        tableView.deselectRow(at: indexPath as IndexPath, animated: true)
        switch((indexPath.item, indexPath.section))
        {
        case (1,0):
            standardValue?.becomeFirstResponder()
            break
        case (0,1):
            customType?.becomeFirstResponder()
            break
        case(1,1):
            customValue?.becomeFirstResponder()
            break
        case(0,0):
            let sheet = IndexedActionSheet.init(title: "Select Standard Action", delegate: self, cancelButtonTitle: "Cancel", destructiveButtonTitle: nil, otherButtonTitles: "dial", "url", "openApp")
            sheet.indexPath = indexPath as NSIndexPath
            sheet.show(in: tableView)
            fallthrough
        default:
            customValue?.resignFirstResponder()
            customType?.resignFirstResponder()
            standardValue?.resignFirstResponder()
            break
        }
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return PushActionItems.count
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int
    {
        return PushActionSections.count
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String?
    {
        if section == 0
        {
            return "Standard Action"
        }
        return "Custom Action"
    }
    
    override func tableView(_ tableView: UITableView, titleForFooterInSection section: Int) -> String?
    {
        
        if section == 0
        {
            return "The action opens a dialer to call the number that is specified in the \"value\" key"
        }
        return "Please look at the application code for the implementation of custom actions (customAction). You can use that code as a starting point to create more custom actions. For more details, please see http://ibm.co/1Fp0OEQ"
    }
    
    func textFieldDidEndEditing(_ textField: UITextField) {
        if let keyedTextField = textField as? KeyedTextField
        {
            UserDefaults.standard.set(keyedTextField.text, forKey: keyedTextField.key!)
            let indexPath = keyedTextField.indexPath!
            let jsonIndexPath = IndexPath(item: 2, section: indexPath.section)
            tableView.reloadRows(at: [indexPath, jsonIndexPath], with: .automatic)
        }
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
    
    func actionSheet(_ actionSheet: UIActionSheet, clickedButtonAt buttonIndex: Int) {
        if let indexedActionSheet = actionSheet as? IndexedActionSheet
        {
            if let pushActionType = PushActionTypes(rawValue: buttonIndex)
            {
                switch(pushActionType)
                {
                case .dial:
                    UserDefaults.standard.set("dial", forKey: "standardType")
                    break
                case .url:
                    UserDefaults.standard.set("url", forKey: "standardType")
                    break
                case .openApp:
                    UserDefaults.standard.set("openApp", forKey: "standardType")
                    break
                }
            }
            let indexPath = indexedActionSheet.indexPath!
            let jsonIndexPath = IndexPath(item: 2, section: indexPath.section)
            let valueIndexPath = IndexPath(item: 1, section: indexPath.section)
            tableView.reloadRows(at: [indexPath as IndexPath, jsonIndexPath, valueIndexPath], with: .automatic)
        }
    }
}

class IndexedActionSheet : UIActionSheet
{
    var indexPath: NSIndexPath?
}
